### PR TITLE
OSSM-4565: [DOC] Add Kiali Configuration for user-workload monitoring

### PR DESCRIPTION
**Pull request by gwynnemonahan**
[OSSM-4565](https://issues.redhat.com//browse/OSSM-4565) [DOC] Add Kiali Configuration for user-workload monitoring

Version(s):
4.10+

Issue:
https://issues.redhat.com/browse/OSSM-4565

Link to docs preview:
https://62902--docspreview.netlify.app/openshift-enterprise/latest/service_mesh/v2x/ossm-observability.html#ossm-integrating-with-user-workload-monitoring_observability

QE review:
- [x] QE has approved this change.

Additional information:
Adding Kiali configuration information for user-workload monitoring as it got left out with initial doc. 

Kiali configuration information are now Steps 1 and 2.

### CODE REVIEW

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 1]
```diff
@@ -15,9 +15,40 @@ The following steps show how to integrate Service Mesh with user-workload monito
 
 * User-workload monitoring is enabled.
 * {SMProductName} Operator 2.4 is installed.
+* Kiali Operator 1.65 is installed.
 
 .Procedure
 
+. Configure Kiali for `user-workload-monitoring`:
```
Comment (jewertow) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:22]:
```suggestion
. Configure Kiali for `user-workload monitoring`:
```

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 2]
```diff
@@ -15,9 +15,40 @@ The following steps show how to integrate Service Mesh with user-workload monito
...
++
+[source,yaml]
+----
+apiVersion: kiali.io/v1alpha1
+kind: Kiali
+metadata:
+  name: kiali-user-workload-monitoring
+  namespace: istio-system-1
```
Comment (jewertow) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:30]:
```suggestion
  namespace: istio-system
```

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 3]
```diff
@@ -15,9 +15,40 @@ The following steps show how to integrate Service Mesh with user-workload monito
...
+spec:
+  external_services:
+    istio:
+      config_map_name: istio-basic
+      istio_sidecar_injector_config_map_name: istio-sidecar-injector-basic
+      istiod_deployment_name: istiod-basic
+      istiod_pod_monitoring_port: 15014
+      url_service_version: 'http://istiod-basic.istio-system-1:15014/version'
```
Comment (jewertow) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:38]:
```suggestion
      url_service_version: 'http://istiod-basic.istio-system:15014/version'
```

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 4]
```diff
@@ -15,9 +15,40 @@ The following steps show how to integrate Service Mesh with user-workload monito
...
+    prometheus:
+      auth:
+        token: secret:thanos-querier-web-token:token
+        type: bearer
+        use_kiali_token: false
+      query_scope:
+        mesh_id: "basic_istio-system-1"
```
Comment (jewertow) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:45]:
```suggestion
        mesh_id: "basic_istio-system"
```
Btw. now I see that we should change `replacement: "<smcp_name>-<smcp_namespace>" # <2>` in Pod/Service monitors and explain in the <2> that this value could be anything unique, e.g. `<smcp_name>-<smcp_namespace>`.

**Review by jewertow — COMMENTED** [review]
We also have to explain how to create a token to Thanos for Kiali:
```
SECRET=`oc get secret -n openshift-user-workload-monitoring | grep  prometheus-user-workload-token | head -n 1 | awk '{print $1 }'`
TOKEN=`echo $(oc get secret $SECRET -n openshift-user-workload-monitoring -o json | jq -r '.data.token') | base64 -d`
oc create secret generic thanos-querier-web-token -n istio-system --from-literal=token=$TOKEN
```

**Review by jewertow — CHANGES_REQUESTED** [review]

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 5]
```diff
@@ -149,7 +199,7 @@ spec:
       action: replace
       targetLabel: pod_name
     - action: replace
-      replacement: "<smcp_name>-<smcp_namespace>" # <2>
+      replacement: <smcp_name>-<smcp_namespace> # <2>
```
[the remark this answers is not in the document]
Comment (gwynnemonahan) [reply modules/ossm-integrating-with-user-workload-monitoring.adoc]: Does "<smcp-name>-<smcp-namespace>" get replaced with "basic-istio-system" in both places? The first comment shows "basic-istio-system" but this comment you have it changing to `<smcp-name>-<smcp-namespace>` and not "basic-istio-system" so wanted to check.

[the remark this answers is not in the document]
Comment (jewertow) [reply modules/ossm-integrating-with-user-workload-monitoring.adoc]: > Does <smcp_name>-<smcp_namespace> get replaced with "basic-istio-system" in both places?

Yes.

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 6]
```diff
@@ -149,10 +199,11 @@ spec:
...
       targetLabel: pod_name
     - action: replace
-      replacement: "<smcp_name>-<smcp_namespace>" # <2>
-      targetLabel: mesh_id
+      replacement: "basic-istio-system" # <2>
+      targetLabel: mesh_id # <3>
 ----
 <1> Since {product-title} monitoring ignores the `namespaceSelector` spec in `ServiceMonitor` and `PodMonitor` objects, you must apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.
-<2> A shared Prometheus instance can include metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the query scope in Kiali to only the relevant mesh metrics.
+<2> Replace `"basic-istio-system"` with a unique value, such as `<smcp-name>-<smcp-namespace>` to create a unique combination within a cluster with multiple SMCPs.
```
Comment (jewertow) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:206]:
```suggestion
<2> `"basic-istio-system"` is a combination of the SMCP name and namespace. It can be replaced with any other unique value within a cluster with multiple SMCPs. This is required only when Kiali is used and multiple SMCPs are deployed in a cluster. In case of cluster-wide deployment, it can be skipped.
```

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 7]
```diff
@@ -15,9 +15,58 @@ The following steps show how to integrate Service Mesh with user-workload monito
...
+      istiod_deployment_name: istiod-basic
+      istiod_pod_monitoring_port: 15014
+      url_service_version: 'http://istiod-basic.istio-system:15014/version'
+    prometheus:
+      auth:
+        token: secret:thanos-querier-web-token:token
+        type: bearer
+        use_kiali_token: false
+      query_scope:
+        mesh_id: "basic_istio-system"
```
Comment (jewertow) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:63]: It must match values from PodMonitor:
```suggestion
        mesh_id: "basic-istio-system"
```

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 8]
```diff
@@ -99,11 +148,12 @@ spec:
...
     relabelings:
     - action: replace
-      replacement: "<smcp_name>-<smcp_namespace>" # <2>
-      targetLabel: mesh_id
+      replacement: "basic-istio-system" # <2>
+      targetLabel: mesh_id # <3>
 ----
 <1> Since {product-title} monitoring ignores the `namespaceSelector` spec in `ServiceMonitor` and `PodMonitor` objects, you must apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.
-<2> A shared Prometheus instance can include metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the query scope in Kiali to only the relevant mesh metrics.
+<2> Since `"basic-istio-system"` is a combination of the SMCP name and namespace, you can replace it with any other unique value within a cluster with multiple SMCPs. You must replace `"basic-istio-system"` if you are using Kiali and have multiple SMCPs deployed in a cluster. You do not need to do this if you have a cluster-wide deployment.
```
Comment (jewertow) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:155]: 1) "You must replace `"basic-istio-system"` if you are using Kiali and have multiple SMCPs deployed in a cluster." - Replace with what? I would prefer to just say that the value of `mesh_id` label must be unique for every mesh in the cluster.
2) "You do not need to do this if you have a cluster-wide deployment." - this sentence is very unclear. What don't I have to do? Changing "basic-istio-system" or adding it at all? I wanted to explain that `mesh_id` label is not necessary in cluster-wide deployments and then users don't need it in Pod/ServiceMonitor as well as in Kiali. In cluster-wide deployments users can remove this replacement:
```
    - action: replace
      replacement: "<smcp_name>-<smcp_namespace>" # <2>
      targetLabel: mesh_id
```
We can explain 2. in <3>.

**Comment by gwynnemonahan** [comment]
@jewertow , do users have to install Kiali in order to integrate with user-workload monitoring?

**Review by jewertow — APPROVED** [review]
LGTM

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 9]
```diff
@@ -15,9 +15,58 @@ The following steps show how to integrate Service Mesh with user-workload monito
...
++
+[source,terminal]
+----
+$ SECRET=`oc get secret -n openshift-user-workload-monitoring | 
+ grep  prometheus-user-workload-token | head -n 1 | awk '{print $1 }'`
+----
++
+[source,terminal]
+----
+$ TOKEN=`echo $(oc get secret $SECRET -n openshift-user-workload-monitoring -o jsonpath='{.data.token}' | base64 -d`
```
Comment (jewertow) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:32]:
```suggestion
$ TOKEN=`echo $(oc get secret $SECRET -n openshift-user-workload-monitoring -o jsonpath='{.data.token}') | base64 -d`
```
It would be great if we could remove `$`, because it does not allow to copy the line and paste in the terminal as is. Instead, we have to copy and remove first character. This is so annoying... Is it necessary?

**Review by jewertow — CHANGES_REQUESTED** [review]
You have to add the following settings to SMCP:
```
spec:
  addons:
    kiali:
      name: kiali-user-workload-monitoring
```

**Comment by jewertow** [comment]
> do users have to install Kiali in order to integrate with user-workload monitoring?

Unfortunately, yes... At least for now.

File: modules/ossm-integrating-with-user-workload-monitoring.adoc (hunk 9 above)
Comment (gwynnemonahan) [reply modules/ossm-integrating-with-user-workload-monitoring.adoc:32]: Yes, unfortunately, the `$` has to remain per Red Hat style guidelines:

https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

If the user must run a command as root, use a number sign (#) at the start of the command instead of a dollar sign ($). For example:

# subscription-manager list

Comment (bburt-rh) [reply modules/ossm-integrating-with-user-workload-monitoring.adoc:32]: @jewertow and @gwynnemonahan - FYI, when a user clicks the Copy icon next to sample code in the docs, the prompt is not included in the copied text:

image

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 10]
```diff
@@ -99,11 +150,11 @@ spec:
...
 ----
 <1> Since {product-title} monitoring ignores the `namespaceSelector` spec in `ServiceMonitor` and `PodMonitor` objects, you must apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.
-<2> A shared Prometheus instance can include metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the query scope in Kiali to only the relevant mesh metrics.
+<2> The variable `"basic-istio-system"` is a combination of the SMCP name and namespace. It can be replaced with any other unique value within a cluster with multiple SMCPs. This is required only when Kiali is used and multiple SMCPs are deployed in a cluster. In case of cluster-wide deployment, it can be skipped. A shared Prometheus instance can include metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the query scope in Kiali to only the relevant mesh metrics.
```
Comment (monroe-dylan-m) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:157]:
```suggestion
<2> Adding the `mesh_id` label can be skipped if there will never be more than one mesh being used with user workload monitoring on the cluster (be sure to also remove the `spec.prometheus.query_scope` from the Kiali deployment if following this guide). Otherwise, adding a `mesh_id` label is recommended as it allows for scoping metrics to their mesh of origin, and for Kiali it is required. The string `"basic-istio-system"` is a combination of the SMCP name and its namespace, but any label can be used as long as it is unique for every mesh using user workload monitoring in the cluster.
```

I took a stab at re-wording this to fix the following issues:

- For clarity, I think `"basic-istio-system"` should be called a "string" not a "variable".
- Additionally, the connection is never made between this value and the `spec.prometheus.query_scope` of the Kiali example resource given above, so it gives the impression that if I only have one mesh in my cluster I can just leave out this `mesh_id` relabeling, when I would also have to clear the `spec.prometheus.query_scope` of the Kiali resource.
- It also misses the nuance that the issue doesn't come from having more than one mesh _on the cluster_, in and of itself. There is only a need for the `mesh_id` label when there is more than one mesh being _who's metrics are being collected by user workload monitoring stack_. If including this detail makes the footnote too long or hard to follow, I don't think it's an important enough distinction to justify itself.

**Review by monroe-dylan-m — CHANGES_REQUESTED** [review]

File: modules/ossm-integrating-with-user-workload-monitoring.adoc (hunk 10 above)
Comment (gwynnemonahan) [reply modules/ossm-integrating-with-user-workload-monitoring.adoc]: Taking your suggestions via Slack of a shorter footnote, and then including the other information as a NOTE. I think that will be easier to follow than cramming it all into a footnote, or trying to separate it all into a different module, esp since it is all part of the same step.

**Review by jewertow — APPROVED** [review]
LGTM

**Comment by rh-tokeefe** [comment]
LGTM

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 11]
```diff
@@ -99,11 +150,18 @@ spec:
...
+<2> The string `"basic-istio-system"` is a combination of the SMCP name and its namespace, but any label can be used as long as it is unique for every mesh using user workload monitoring in the cluster. The `spec.prometheus.query_scope` of the Kiali resource configured in Step 2 needs to match this value.
++
+[NOTE]
+====
+If there is only one mesh using user-workload monitoring, then both the `mesh_id` relabeling and the `spec.prometheus.query_scope` field in the Kiali resource are optional (but the `query_scope` field given here should be removed if the `mesh_id` label is removed).
+
+If there may be multiple meshes using user-workload monitoring, then both the `mesh_id` relabelings and the `spec.prometheus.query_scope` field in the Kiali resource are required so that Kiali only sees metrics from its associated mesh. If Kiali is not being deployed, applying the `mesh_id` relabeling is still recommended so that metrics from different meshes can be distinguished from one another.
```
Comment (jeana-redhat) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:163]: s/may/might per IBM SG

File: modules/ossm-integrating-with-user-workload-monitoring.adoc [hunk 12]
```diff
@@ -149,10 +207,17 @@ spec:
...
```
Comment (jeana-redhat) [inline modules/ossm-integrating-with-user-workload-monitoring.adoc:220]: s/may/might

**Review by jeana-redhat — COMMENTED** [review]
Two commits must be squashed to one. Also noticed a couple small things to look at.

/remove-label merge-review-in-progress
/remove-label merge-review-needed

**Comment by gwynnemonahan** [comment]
/label merge-review-needed

Doc peer review comments have been resolved and changes made.
